compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/GBKTransform.java

```diff
@@ -0,0 +1,313 @@
+          .removeIf(entry -> entry.getValue().getTimestamp() == minWatermarkTimestamp);
+      }
+
+      minWatermarkHold = keyOutputWatermarkMap.isEmpty()
+        ? new Watermark(Long.MAX_VALUE) : Collections.min(keyOutputWatermarkMap.values());
+
+      outputWatermarkCandidate = new Watermark(
+        Math.max(prevOutputWatermark.getTimestamp(),
+          Math.min(minWatermarkHold.getTimestamp(), inputWatermark.getTimestamp())));
+    }
+  }
+
+  /**
+   * Trigger timers that need to be fired at {@param watermark}.
+   * @param watermark watermark
+   */
+  @Override
+  public void onWatermark(final Watermark watermark) throws RuntimeException {
+    if (watermark.getTimestamp() <= inputWatermark.getTimestamp()) {
+      throw new RuntimeException("Received watermark is before inputWatermark in GBKTransform");
+    }
+    checkAndInvokeBundle();
+    inputWatermark = watermark;
+    // Triggering timers
+    try {
+      processElementsAndTriggerTimers(Instant.now(), Instant.now(), inputWatermark);
+    } catch (final Exception e) {
+      e.printStackTrace();
+      throw new RuntimeException(e);
+    }
+    checkAndFinishBundle();
+  }
+
+  /**
+   * This advances the input watermark and processing time to the timestamp max value
+   * in order to emit all data.
+   */
+  @Override
+  protected void beforeClose() {
+    // Finish any pending windows by advancing the input watermark to infinity.
+    inputWatermark = new Watermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+    processElementsAndTriggerTimers(
+      BoundedWindow.TIMESTAMP_MAX_VALUE, BoundedWindow.TIMESTAMP_MAX_VALUE, inputWatermark);
+  }
+
+  /**
+   * Trigger timers. When triggering, it emits the windowed data to downstream operators.
+   * @param processingTime processing time
+   * @param synchronizedTime synchronized time
+   * @param watermark watermark
+   */
+  private void triggerTimers(final Instant processingTime,
+                            final Instant synchronizedTime,
+                            final Watermark watermark) {
+
+    Iterator<Map.Entry<K, InMemoryTimerInternals>> iter =
+      inMemoryTimerInternalsFactory.getTimerInternalsMap().entrySet().iterator();
+    while (iter.hasNext()) {
+      final Map.Entry<K, InMemoryTimerInternals> curr = iter.next();
+      try {
+        curr.getValue().advanceInputWatermark(new Instant(watermark.getTimestamp()));
+        curr.getValue().advanceProcessingTime(processingTime);
+        curr.getValue().advanceSynchronizedProcessingTime(synchronizedTime);
+      } catch (final Exception e) {
+        e.printStackTrace();
+        throw new RuntimeException();
+      }
+      for (TimeDomain domain : TimeDomain.values()) {
+        processTrigger(curr.getKey(), curr.getValue(), domain);
+      }
+      // Remove timerInternals and stateInternals that are no longer needed.
+      if (inMemoryTimerInternalsFactory.isEmpty(curr.getValue())) {
+        iter.remove();
```

Comment:
Don't we remove stateInternals?

Reply:
Thank you for pointing it out. I added the new line to remove stateInternals.